### PR TITLE
coll: adding various features to gentran

### PR DESCRIPTION
**Pull request by zhenggb72**
A collection of commits to add new features including new vertex types 

1-3:   adding ibcast error handling in gentran infrastructure and collective algorithms
4-7:   various fixes 
8-9:   optimization in gentran for multicast
10:  adding sub-scheduler in gentran, can be used for composition, pipelining for example

They can potentially be split into 4 smaller PRs, however, they have to go in the same order anyway because of the dependences.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by zhenggb72** [comment]
please review

**Comment by hzhou** [comment]
@wesbland @zhenggb72 To expedite the review, consider splitting the PRs into smaller easier pieces. I suggest find trivia commits and get them reviewed first. By trivia, I mean both the motivation is obvious and the changes are not complicated. Commits with either non-obvious (or justifiable) motivations or debatable changes are better to be reviewed one at a time. And one can help the review by supply sufficient justifications.

**Comment by zhenggb72** [comment]
> @wesbland @zhenggb72 To expedite the review, consider splitting the PRs into smaller easier pieces. I suggest find trivial commits and get them reviewed first. By trivial, I mean both the motivation is obvious and the changes are not complicated. Commits with either non-obvious (or justifiable) motivations or debatable changes are better to be reviewed one at a time. And one can help the review by supply sufficient justifications.

Can you review the rest of commits now? They are mostly related to adding new vertex types and features.

**Comment by zhenggb72** [comment]
please review

File: src/mpi/coll/transports/gentran/gentran_types.h [hunk 1]
```diff
@@ -113,6 +129,7 @@ typedef struct MPII_Genutil_sched_t {
     int completed_vtcs;
     int last_fence;
 
+    MPIR_Request *req;
```
Comment (hzhou) [inline src/mpi/coll/transports/gentran/gentran_types.h:132]: This addition is not described in the commit message. Some rationale would be helpful.

File: src/mpi/coll/transports/gentran/tsp_gentran.c [hunk 2]
```diff
@@ -95,8 +95,20 @@ int MPIR_TSP_sched_new_type(MPIR_TSP_sched_t s, MPIR_TSP_sched_issue_fn issue_fn
...
+
+    /* only free_fn can be NULL */
+    MPIR_Assert(issue_fn && complete_fn);
+
+    /* search if the new type already exists in the table */
+    vtype = ut_type_array(&sched->generic_types, MPII_Genutil_vtx_type_t *);
+    for (i = 0; i < utarray_len(&sched->generic_types); i++) {
+        if (vtype->issue_fn == issue_fn && vtype->complete_fn == complete_fn &&
+            vtype->free_fn == free_fn)
+            return MPII_GENUTIL_VTX_KIND__LAST + i + 1;
```
Comment (hzhou) [inline src/mpi/coll/transports/gentran/tsp_gentran.c:147]: It's a minor nitpick -- it is preferable to have explicit braces especially when either the condition or statement resulted in multiple lines.

File: src/mpi/coll/cvars.txt [hunk 3]
```diff
@@ -170,6 +170,17 @@ cvars:
         nb                      - Force nonblocking algorithm
         remote_send_local_bcast - Force remote-send-local-bcast algorithm
 
+    - name        : MPIR_CVAR_IBCAST_ERROR_CHECKING
```
Comment (hzhou) [inline src/mpi/coll/cvars.txt:173]: This cvar is neither general nor specific. It doesn't apply to all IBCAST algorithms, so it is a wrong name if view it as specific. It singles out ibcast yet I don't see the rationale for singling out ibcast -- usually when user desire error checking, they desire it for all collectives.

How about we just use the macro `HAVE_ERROR_CHECKING`?

File: src/mpi/coll/iallgather/iallgather_tsp_brucks.c [hunk 4]
```diff
@@ -84,7 +84,8 @@ MPIR_TSP_Iallgather_sched_intra_brucks(const void *sendbuf, MPI_Aint sendcount,
                                  recvcount, recvtype, sched, 0, NULL);
 
     /* All following sends/recvs and copies depend on this dtcopy */
-    MPIR_TSP_sched_fence(sched);
+    mpi_errno = MPIR_TSP_sched_fence(sched);
+    MPIR_ERR_CHECK(mpi_errno);
```
Comment (hzhou) [inline src/mpi/coll/iallgather/iallgather_tsp_brucks.c:88]: Shouldn't we use the `MPIR_ERR_COLL_CHECKANDCONT` macro as in https://github.com/pmodels/mpich/pull/5491?

Actually, are you sure this commit is not covered in #5491 already?

File: src/mpi/coll/transports/gentran/gentran_utils.c [hunk 5]
```diff
@@ -24,8 +24,9 @@ static inline void vtx_record_issue(MPII_Genutil_vtx_t * vtxp, MPII_Genutil_sche
     LL_APPEND(sched->issued_head, sched->issued_tail, vtxp);
 }
 
-static void vtx_issue(int vtxid, MPII_Genutil_vtx_t * vtxp, MPII_Genutil_sched_t * sched)
+static int vtx_issue(int vtxid, MPII_Genutil_vtx_t * vtxp, MPII_Genutil_sched_t * sched)
```
Comment (hzhou) [inline src/mpi/coll/transports/gentran/gentran_utils.c:27]: The commit message `coll: fix a warning for gentran ` is not good since it gives no clue what this commit changes. I think if you simply swap the short and long messages it would be much better.

File: src/include/mpir_request.h [hunk 6]
```diff
@@ -215,6 +215,7 @@ struct MPIR_Request {
 #endif                          /* HAVE_DEBUGGER_SUPPORT */
             /* Persistent requests have their own "real" requests */
             struct MPIR_Request *real_request;
+            MPIR_TSP_sched_t sched;
```
Comment (hzhou) [inline src/include/mpir_request.h:218]: Can't review this until the use case -- the composition of schedulers -- is explained and understood.

File: src/mpi/coll/transports/gentran/gentran_utils.c [hunk 7]
```diff
@@ -341,17 +341,17 @@ int MPII_Genutil_progress_hook(int *made_progress)
     in_genutil_progress = 1;
 
     if (made_progress)
-        *made_progress = FALSE;
+        *made_progress = false;
```
Comment (hzhou) [inline src/mpi/coll/transports/gentran/gentran_utils.c:345]: Please revise commit messages (` coll/tsp: only count towards PROGRESS_MAX_COLLS when actually makes p…

…rogress`). The short message breaking up is no good. It didn't explain the addition of local progress variable (no matter how obvious it appears at the time).

File: src/mpi/coll/transports/gentran/gentran_impl.c [hunk 8]
```diff
@@ -10,7 +10,7 @@
     - name        : MPIR_CVAR_PROGRESS_MAX_COLLS
       category    : COLLECTIVE
       type        : int
-      default     : 8
+      default     : 0
```
Comment (hzhou) [inline src/mpi/coll/transports/gentran/gentran_impl.c:13]: Please add in the commit message its rationale -- performance impact?

**Review by hzhou — COMMENTED** [review]
There are quite a few different subjects in this PR -- error checking, new vtx types, sub schedules, etc. It appears they are independent changes. If you split them into smaller, more focused PRs, it will make review easier and expedite the process.

The PR title -- "Gentran commits" -- is bad, right? It doesn't raise itself to the review priority. 

cc @wesbland

File: src/mpi/coll/transports/gentran/gentran_types.h (hunk 1 above)
Comment (zhenggb72) [reply src/mpi/coll/transports/gentran/gentran_types.h:132]: It is used in this same commit, in function vtx_issue. This is to support ibcast error propagation (e.g. along broadcast tree). The error is saved in req->u.nbc.errflag. so a pointer to the MPI_Request is added here. If I remember correctly, the MPIDU_Sched also store MPI_Request for the same reason for ibcast error checking.

File: src/mpi/coll/transports/gentran/tsp_gentran.c (hunk 2 above)
Comment (zhenggb72) [reply src/mpi/coll/transports/gentran/tsp_gentran.c:147]: will fix

File: src/mpi/coll/cvars.txt (hunk 3 above)
Comment (zhenggb72) [reply src/mpi/coll/cvars.txt:173]: I think we had a debate on that a long time back. I will have to recall the reason for this. I think It has something to with the MPI standard, since MPI runtime is required to report error in some cases in ibcast. So we prefer not to eliminate it by using the macro. With the specific env var,  the user basically specify he/she is ok with the non-standard behavior for performance reason.

File: src/mpi/coll/transports/gentran/gentran_impl.c (hunk 8 above)
Comment (zhenggb72) [reply src/mpi/coll/transports/gentran/gentran_impl.c]: done, but moved to a new PR at: https://github.com/pmodels/mpich/pull/5506

File: src/mpi/coll/transports/gentran/gentran_utils.c (hunk 7 above)
Comment (zhenggb72) [reply src/mpi/coll/transports/gentran/gentran_utils.c]: I am not sure I get it. But I edited the commit message in the other PR.

File: src/include/mpir_request.h (hunk 6 above)
Comment (zhenggb72) [reply src/include/mpir_request.h:218]: This commit basically adds a new type of vertex in tsp, which is another scheduler. So this basically allows recursively adding schedulers to an existing schedule. Let me know if this general explanation is  acceptable. If not, I will see I will remove this commit for now.

File: src/mpi/coll/cvars.txt (hunk 3 above)
Comment (zhenggb72) [reply src/mpi/coll/cvars.txt]: You were right. I found we have a later commit actually removed the CVAR. I squashed it to this commit now, so it is gone,

File: src/mpi/coll/iallgather/iallgather_tsp_brucks.c (hunk 4 above)
Comment (zhenggb72) [reply src/mpi/coll/iallgather/iallgather_tsp_brucks.c]: I didn't know we have a new macro for error checking. I updated the commit to switch to it,

File: src/mpi/coll/transports/gentran/gentran_utils.c (hunk 5 above)
Comment (zhenggb72) [reply src/mpi/coll/transports/gentran/gentran_utils.c:27]: fixed commit message

**Comment by zhenggb72** [comment]
I believe I addressed all the review comments above. 
Let me know if this PR is still too big, I will try to split it further.

File: src/mpi/coll/transports/gentran/tsp_gentran.c [hunk 9]
```diff
@@ -95,8 +95,20 @@ int MPIR_TSP_sched_new_type(MPIR_TSP_sched_t s, MPIR_TSP_sched_issue_fn issue_fn
                             MPIR_TSP_sched_complete_fn complete_fn, MPIR_TSP_sched_free_fn free_fn)
 {
     MPII_Genutil_sched_t *sched = s;
-    MPII_Genutil_vtx_type_t *newtype;
-    int type_id;
+    MPII_Genutil_vtx_type_t *vtype, *newtype;
+    int type_id, i;
+
+    /* only free_fn can be NULL */
+    MPIR_Assert(issue_fn && complete_fn);
```
Comment (raffenet) [inline src/mpi/coll/transports/gentran/tsp_gentran.c:140]: this seems unrelated to this commit

File: src/mpi/coll/transports/gentran/tsp_gentran.c [hunk 10]
```diff
@@ -331,6 +364,28 @@ int MPIR_TSP_sched_localcopy(const void *sendbuf, MPI_Aint sendcount, MPI_Dataty
...
+
+    /* record the arguments */
+    vtxp->u.cb.cb_p = cb_p;
+    vtxp->u.cb.cb_data = cb_data;
+
+    MPL_DBG_MSG_FMT(MPIR_DBG_COLL, VERBOSE, (MPL_DBG_FDEST, "Gentran: schedule [%d] cb", vtx_id));
+
+    return vtx_id;
+}
+
```
Comment (raffenet) [inline src/mpi/coll/transports/gentran/tsp_gentran.c:452]: this is probably not the time for it, but wouldn't it be possible to implement all vertex callbacks using this function signature? just the stick the arguments in `void *cb_data`.

File: src/mpi/coll/transports/gentran/gentran_types.h (hunk 1 above)
Comment (raffenet) [reply src/mpi/coll/transports/gentran/gentran_types.h:132]: The next commit in the series adds error checking to ibcast. From what you are saying, this change belongs to that commit, not this one.

File: src/mpi/coll/ibcast/ibcast_tsp_tree.c [hunk 11]
```diff
@@ -50,19 +51,38 @@ int MPIR_TSP_Ibcast_sched_intra_tree(void *buffer, MPI_Aint count, MPI_Datatype
...
+        struct MPII_Ibcast_state *ibcast_state = NULL;
+#endif
         int msgsize = (i == 0) ? chunk_size_floor : chunk_size_ceil;
 
+#ifdef HAVE_ERROR_CHECKING
+        ibcast_state = MPIR_TSP_sched_malloc(sizeof(struct MPII_Ibcast_state), sched);
+        if (ibcast_state == NULL)
+            MPIR_ERR_POP(mpi_errno);
+        ibcast_state->n_bytes = msgsize * type_size;
+#endif
```
Comment (raffenet) [inline src/mpi/coll/ibcast/ibcast_tsp_tree.c:64]: Combine these 2 sections into one #ifdef.

File: src/mpi/coll/ibcast/ibcast_tsp_tree.c [hunk 12]
```diff
@@ -50,19 +51,38 @@ int MPIR_TSP_Ibcast_sched_intra_tree(void *buffer, MPI_Aint count, MPI_Datatype
...
+                                            datatype, my_tree.parent, tag, comm,
+                                            &ibcast_state->status, sched, 0, NULL, &recv_id);
+            MPIR_ERR_COLL_CHECKANDCONT(mpi_errno, errflag);
+            MPIR_TSP_sched_cb(&MPII_Ibcast_sched_test_length, ibcast_state, sched, 1, &recv_id);
+#else
             mpi_errno =
                 MPIR_TSP_sched_irecv((char *) buffer + offset * extent, msgsize, datatype,
                                      my_tree.parent, tag, comm, sched, 0, NULL, &recv_id);
             MPIR_ERR_COLL_CHECKANDCONT(mpi_errno, errflag);
+#endif
```
Comment (raffenet) [inline src/mpi/coll/ibcast/ibcast_tsp_tree.c:85]: If `MPIR_TSP_sched_irecv_status` supports `MPI_STATUS_IGNORE`, is there any reason not to always use it?

File: src/mpi/coll/transports/gentran/tsp_gentran.c [hunk 13]
```diff
@@ -557,6 +562,7 @@ int MPIR_TSP_sched_sub_sched(MPIR_TSP_sched_t s, MPIR_TSP_sched_t subs,
     MPII_Genutil_vtx_add_dependencies(sched, *vtx_id, n_in_vtcs, in_vtcs);
     vtxp->u.sched.sched = subsched;
     vtxp->u.sched.req = NULL;
+    vtxp->u.sched.is_persistent = subsched->is_persistent;
```
Comment (raffenet) [inline src/mpi/coll/transports/gentran/tsp_gentran.c:566]: I think this and the previous commit make more sense squashed into the commit that adds the sub-scheduler vertex type.

File: src/mpi/coll/transports/gentran/tsp_gentran.c (hunk 9 above)
Comment (zhenggb72) [reply src/mpi/coll/transports/gentran/tsp_gentran.c:140]: Sure, should I make a new commit for that?

File: src/mpi/coll/transports/gentran/tsp_gentran.c (hunk 10 above)
Comment (zhenggb72) [reply src/mpi/coll/transports/gentran/tsp_gentran.c:452]: Can you please elaborate on that?
This function only adds a callback vertex into gentran scheduler. It does not actually call the callback. All callback functions are required to have  the same signature (MPIR_TSP_cb_t).

File: src/mpi/coll/transports/gentran/tsp_gentran.c (hunk 9 above)
Comment (raffenet) [reply src/mpi/coll/transports/gentran/tsp_gentran.c:140]: sure

File: src/mpi/coll/ibcast/ibcast_tsp_tree.c (hunk 11 above)
Comment (zhenggb72) [reply src/mpi/coll/ibcast/ibcast_tsp_tree.c:64]: will fix

File: src/mpi/coll/ibcast/ibcast_tsp_tree.c (hunk 12 above)
Comment (zhenggb72) [reply src/mpi/coll/ibcast/ibcast_tsp_tree.c:85]: Maybe I missed something, but for ibcast, user does not pass a status argument.

File: src/mpi/coll/transports/gentran/tsp_gentran.c (hunk 10 above)
Comment (raffenet) [reply src/mpi/coll/transports/gentran/tsp_gentran.c:452]: What I am trying to say is that `MPIR_TSP_sched_irecv_status` and `MPIR_TSP_sched_cb` are basically the same function. Perhaps all vertexes could be added as generic callbacks. Anyway this is way out of scope for this PR, so lets just drop it.

File: src/mpi/coll/transports/gentran/tsp_gentran.c (hunk 13 above)
Comment (zhenggb72) [reply src/mpi/coll/transports/gentran/tsp_gentran.c:566]: The previous change adding "req" to scheduler is part of that commit because the new vertex type added (for error checking) uses the "req" in the same commit. Moving req to later commit breaks it. I think even the commit to add schede_irecv_status technically speaking is also part of the error checking commit group.

And this change - propagating the is_persistent from a scheduler to the sub-scheduler vertex in the parent tree is also part of this commit, because otherwise, freeing the parent scheduler can accidentally freeing the subscheduler which is a persistent scheduler. This is exactly what this commit was trying to fix.

File: src/mpi/coll/ibcast/ibcast_tsp_tree.c (hunk 12 above)
Comment (raffenet) [reply src/mpi/coll/ibcast/ibcast_tsp_tree.c:83]: if no error checking, set `ibcast_state->status = MPI_STATUS_IGNORE`.

Comment (raffenet) [reply src/mpi/coll/ibcast/ibcast_tsp_tree.c:83]: maybe something like this
```c
            MPI_Status *status = MPI_STATUS_IGNORE;
#ifdef HAVE_ERROR_CHECKING
            status = &ibcast_state->status;
#endif
            mpi_errno =
                MPIR_TSP_sched_irecv_status((char *) buffer + offset * extent, msgsize,
                                            datatype, my_tree.parent, tag, comm,
                                            status, sched, 0, NULL, &recv_id);
            MPIR_ERR_COLL_CHECKANDCONT(mpi_errno, errflag);
            MPIR_TSP_sched_cb(&MPII_Ibcast_sched_test_length, ibcast_state, sched, 1, &recv_id);
```

File: src/mpi/coll/transports/gentran/gentran_types.h (hunk 1 above)
Comment (zhenggb72) [reply src/mpi/coll/transports/gentran/gentran_types.h:132]: Adding "req" to scheduler is part of that commit because the new vertex type added (for error checking) uses the "req" in the same commit. Moving req to later commit breaks it. I think even the commit to add schede_irecv_status technically speaking is also part of the error checking commit group.

File: src/mpi/coll/ibcast/ibcast_tsp_tree.c (hunk 11 above)
Comment (zhenggb72) [reply src/mpi/coll/ibcast/ibcast_tsp_tree.c]: done

File: src/mpi/coll/ibcast/ibcast_tsp_tree.c (hunk 12 above)
Comment (zhenggb72) [reply src/mpi/coll/ibcast/ibcast_tsp_tree.c:83]: I see, in that case, we don't need to have two separate vertices (sched_irecv and sched_irecv_status). They can be combined into one after making it supporting MPI_STATUS_IGNORE.

This sounds good to me. But can we do that in a separate PR?

File: src/mpi/coll/transports/gentran/tsp_gentran.c (hunk 13 above)
Comment (raffenet) [reply src/mpi/coll/transports/gentran/tsp_gentran.c:566]: There are 3 commits
```
coll: add a new Gentran vertex type to encapsulate a sub scheduler
coll/tsp: reset schedule function needs to reset sub-schedule
coll/tsp: free sub scheduler when the main scheduler is freed
```
My understanding is that the first adds a broken sub-schedule vertex that is subsequently fixed up by the next 2 commits. My suggestion is to squash them into a single, non-broken commit. Or is my understanding wrong?

Comment (zhenggb72) [reply src/mpi/coll/transports/gentran/tsp_gentran.c:566]: Got you. Yes, that makes sense.

File: src/mpi/coll/transports/gentran/tsp_gentran.c (hunk 9 above)
Comment (zhenggb72) [reply src/mpi/coll/transports/gentran/tsp_gentran.c:140]: done

File: src/mpi/coll/transports/gentran/tsp_gentran.c (hunk 13 above)
Comment (zhenggb72) [reply src/mpi/coll/transports/gentran/tsp_gentran.c:566]: done

File: src/mpi/coll/ibcast/ibcast_tsp_tree.c (hunk 12 above)
Comment (raffenet) [reply src/mpi/coll/ibcast/ibcast_tsp_tree.c:83]: > This sounds good to me. But can we do that in a separate PR?

Sure.

File: src/mpi/coll/transports/gentran/gentran_types.h (hunk 1 above)
Comment (raffenet) [reply src/mpi/coll/transports/gentran/gentran_types.h:132]: OK, that helps my understanding. It's sloppy commit boundaries like this that make it hard to review PRs, which results in things dragging out for a long time.

I'm not saying you need to go and fix the history this time. But please be aware for future PRs.

**Review by raffenet — APPROVED** [review]

File: src/mpi/coll/transports/gentran/gentran_types.h (hunk 1 above)
Comment (zhenggb72) [reply src/mpi/coll/transports/gentran/gentran_types.h:132]: Thanks for the comments. I understand reviewing this PR is challenging. Hopefully it is just this one. This PR is a collection of commits with pretty old history, which is also complicated by several rounds of code refactorization. 

Could you please also review https://github.com/pmodels/mpich/pull/5506 which was split from this PR.

**Comment by wesbland** [comment]
Looks like this has been approved and just needs a rebase. @zhenggb72 can you take care of that?

**Comment by hzhou** [comment]
@zhenggb72 Please check the Author checklist (actually read and check). Thanks!

**Comment by zhenggb72** [comment]
Do we run any Jenkins test?

**Comment by zhenggb72** [comment]
> Looks like this has been approved and just needs a rebase. @zhenggb72 can you take care of that?

done

**Comment by hzhou** [comment]
> 
> 
> Do we run any Jenkins test?

If it cleared the tests before rebase and the rebase is without complications, then you can skip the extra jenkins tests.

**Comment by hzhou** [comment]
Just in case, I don't need to review this PR again. Since Ken has approved it, it is okay to merge after rebase.

**Comment by zhenggb72** [comment]
rebased, thanks
